### DISCUSSION TITLE
change file watcher to `@parcel/watcher`

### DISCUSSION THREAD

**Discussion by bytemain** [opensumi/core#302] [General]
https://www.npmjs.com/package/@parcel/watcher

- https://github.com/microsoft/vscode/issues/135653

also can add an options:
- https://github.com/microsoft/vscode/issues/135891

**Comment by Aaaaash** [comment]
我们也需要这个 options 吗

**Comment by bytemain** [comment]
vscode 讨论是否要更换一个 fs watcher 的 issue:
https://github.com/microsoft/vscode/issues/132483

**Comment by bytemain** [comment]
各位有遇到什么 nsfw 现在的坑吗？

翻聊天记录看到一个创建文件相关的问题：

具体聊天记录就不贴了，大家在群里搜一下 nsfw，2021-11 月份的讨论。

hacke2:
我新增的是 file:///home/admin/xxxx/kaitian/ide-framework/packages/ide-file-service/src/node/444
```
{"changes":[{"uri":"file:///home/admin/xxxx/kaitian/ide-framework/packages/addons/node_modules/%40ali/ide-editor/node_modules/%40ali/ide-file-service/src/node/444","type":1}]}
{"changes":[{"uri":"file:///home/admin/xxxx/kaitian/ide-framework/packages/addons/node_modules/%40ali/ide-editor/node_modules/%40ali/ide-file-service/src/node/444","type":1}]}
```

我们过滤的是 **/node_module/**，所以上面的变更如果不找到 realpath 就会被过滤掉

最后得出的结论应该就是 nsfw 的问题：

- https://github.com/Axosoft/nsfw/issues/67

**Comment by bytemain** [comment]
小程序 IDE 遇到过一个文件监听的问题：

通过 preference 关联配置项文件，然后疯狂的点击配置项文件的 checkbox，PreferenceProxy 会把修改后的值写入文件中。
但是我监听这个文件改变的时候，发现就有两种情况：
![image](https://user-images.githubusercontent.com/13938334/154413727-c3910dc9-37b1-43f4-b03f-a0668866c865.png)
![image](https://user-images.githubusercontent.com/13938334/154413752-a8dfedfb-5381-4dd5-8a38-182c12626c84.png)


感觉是 FileChangeEvent 有问题

**Reply by erha19** [reply]
一部分原因可能也出在依赖 `write-file-atomic` 实现的文件修改上

**Reply by bytemain** [reply]
> 一部分原因可能也出在依赖 `write-file-atomic` 实现的文件修改上

我觉得你说得对

**Comment by bytemain** [comment]
- https://github.com/microsoft/vscode/commit/85c5eb7281c69c4a83ba923163f0f6f1f6d904a7

vscode 转移到 @parcel/watcher 的 commit

**Comment by bytemain** [comment]
related: https://github.com/opensumi/core/issues/867

**Comment by erha19** [comment]
### nsfw & @parcel/watcher 使用效果对比

结合当前框架内使用 `write-file-atomic` 的写入方案，对比了一下两个库的差异

#### NSFW

![image](https://user-images.githubusercontent.com/9823838/181440600-95fd2634-203d-4bc0-8172-7a0f9bd33c9c.png)

#### 测试脚本

```js
const nsfw = require('nsfw');
const path = require('path');
const writeFile = require('write-file-atomic');

(async () => {
  const watch = await nsfw(path.resolve('./tests'), (event) => {
    console.log('-----');
    console.log(event);
    console.log('-----');
  });
  let content = 1;
  watch.start();
  setInterval(() => {
    content ++;
    const buffer = Buffer.from(Uint8Array.from(content));
    writeFile(path.resolve('./tests/index.js'), buffer);
  }, (2000));
})()
```

#### Mac 环境
结果： 临时文件的  DELETE，以及两次原始文件的 CREATED，事件行为不符合预期
![image](https://user-images.githubusercontent.com/9823838/181441023-db8904b2-ac38-4ea6-ae5d-6421a2542952.png)

#### Windows 环境

结果：临时文件的 CREATED，原文件的 DELETE，临时文件的 RENAME
![image](https://user-images.githubusercontent.com/9823838/181441070-0d5d66f9-479b-49cc-a4cf-5ce06839202a.png)

#### Linux 环境
结果： 稳定的 0，,2，2，,3，即 临时文件的 CREATE，MODIFIED，MODIFIED，RENAME 事件
![image](https://user-images.githubusercontent.com/9823838/181441135-9a732910-5743-4d4b-a141-8d0743d62392.png)


### @parcel/watcher

```js
const watcher = require('@parcel/watcher');
const path = require('path');
const writeFile = require('write-file-atomic');

(async () => {
  const watch = await watcher.subscribe(path.resolve('./tests'), (err, event) => {
    console.log('-----');
    console.log(event);
    console.log('-----');
  });
  let content = 1;
  setInterval(() => {
    content ++;
    const buffer = Buffer.from(Uint8Array.from(content));
    writeFile(path.resolve('./tests/index.js'), buffer);
  }, (2000));
})()
```

#### Mac 环境
结果：得到的结果基本上为单个的 CREATE 事件以及偶现的 临时文件 DELETE
![image](https://user-images.githubusercontent.com/9823838/181441338-b1395b59-7fe0-4db9-baba-4081fafade57.png)

#### Windows 环境
结果：准确的单个 UPDATE 事件
![image](https://user-images.githubusercontent.com/9823838/181441368-7f1ca680-e61b-4c9a-a3c7-5c9712db5ad0.png)

#### Linux 环境
结果：稳定的原文件 CREATE 事件

![image](https://user-images.githubusercontent.com/9823838/181441424-cc51c7a8-f4c5-4ffd-997b-e41761ba07c4.png)


### 最终结论

在 Windows 环境下 `@parcel/watcher` 的表现更加准确，且在 Mac 及 Linux 环境下尽管表现不符预期，但事件单一，不会出现临时文件的冗余事件，表现也更加良好。

**Comment by erha19** [comment]
Done on #1437.

**Comment by songhn233** [comment]
提醒一个额外引入的本地开发问题，在 native 构建时候我当时出现了报错

在一堆 error 中 `env: python: No such file or directory`

`@parcel/watcher` 需要 python，而本地只配置了 python3，其他模块构建时不会出现这个问题

可以通过软链接搞过去 
```sh
 ln -s -f /opt/homebrew/bin/python3 /opt/homebrew/bin/python
```
